Fabrid/base

jeltevanbommel opened the pull request.

matzf committed:
Add Carbon Intensity information to beacons and paths (#129)

* Add Carbon Intensity Information to beacons
First crude version, just to allow adding _some_ numbers.
Note: not merged code paths with bandwidth although this is currently
processed identically, as changes to the representation of the carbon
intensity information are to be expected.

* Add Carbon Intensity to combinator, snet.Path and showpaths

* Resolved conflict in daemon protobuf file by incrementing field
  number. Recreated protobuf generated file.
* Resolved conflicts in other places due to diff not matching original,
  by adding the carbon intentity field in data
structures/initializations again to the upstream ones.

matzf committed:
Add Carbon Intensity information to beacons and paths (#129)

* Add Carbon Intensity Information to beacons
First crude version, just to allow adding _some_ numbers.
Note: not merged code paths with bandwidth although this is currently
processed identically, as changes to the representation of the carbon
intensity information are to be expected.

* Add Carbon Intensity to combinator, snet.Path and showpaths

* Resolved conflict in daemon protobuf file by incrementing field
  number. Recreated protobuf generated file.
* Resolved conflicts in other places due to diff not matching original,
  by adding the carbon intentity field in data
structures/initializations again to the upstream ones.

rohrerj committed:
fabrid hbh extension and unit tests

rohrerj committed:
updated hbh header approach

rohrerj committed:
started with fabrid crypto implementation

rohrerj committed:
added tests for identifier serialization

rohrerj committed:
fabrid crypto functions and unit tests

rohrerj committed:
fix e2e to hbh type

rohrerj committed:
started with fabrid router implementation

rohrerj committed:
unit tests and some small updates

rohrerj committed:
add QoS field to fabrid hopfield meta

rohrerj committed:
updated fabrid hvf computation

rohrerj committed:
changed fabrid hbh mac computation from cmac to cbc

rohrerj committed:
changed padding to zero padding

rohrerj committed:
moved fabrid crypto logic to experimental package

matzf committed:
deps: bump google.golang.org/grpc, set MaxConcurrentStreams (#4424)

Bump google.golang.org/grpc from 1.57.0 to 1.57.2 due to a
security patch.
Skipping 1.57.1, because that contained a bug.

From vulnerability alert "gRPC-Go HTTP/2 Rapid Reset vulnerability":

> Impact
>
> In affected releases of gRPC-Go, it is possible for an attacker to send
> HTTP/2 requests, cancel them, and send subsequent requests, which is
> valid by the HTTP/2 protocol, but would cause the gRPC-Go server to
> launch more concurrent method handlers than the configured maximum
> stream limit. Patches
>
> This vulnerability was addressed by #6703 and has been included in patch
> releases: 1.56.3, 1.57.1, 1.58.3. It is also included in the latest
> release, 1.59.0.
>
> Along with applying the patch, users should also ensure they are using
> the grpc.MaxConcurrentStreams server option to apply a limit to the
> server's resources used for any single connection.

jiceathome committed:
router: extend packet processing metrics (#4422)

Improve metrics for router packet processing:

* Classify packets by size bucket: log2(size)
* Classify packets by traffic type: incoming, outgoing, as-transit-in/out, br-transit.
* Add process-level on-cpu and sleep time accounting

oncilla committed:
build: update Go to 1.21.3 (#4428)

oncilla committed:
build: drop @cgrindel_bazel_starlib//updatesrc (#4427)

Refactor the remaining targets that depended on @cgrindel_bazel_starlib//updatesrc.
The repository has now been completely migrated to @aspect_bazel_lib//lib:write_source_files.

During the refactor, we dropped the functionality to write the source files in
rules_openapi. Instead, the caller side can decide how to write the source
files. This makes the rules more composable and easier to use. The functionality
has been moved to a macro in private/mgmtapi/api.bzl, which is used inside of
scionproto/scion.

lukedirtwalker committed:
deps: update bazel & bazel libs (#4426)

Drop the `go_embed_data` custom rules, those were deprecated for a long time.
Drop most of the custom `rules_openapi` bazel rules.
Users are recommended to use their own invocations of the JS tools.

jeltevanbommel committed:
wip control plane

oncilla committed:
build: update quic-go to v0.40.0 (#4431)

jiceathome committed:
router: fix interface label of router metrics (#4430)

The interface label value of every metric instance had been lost to a refactoring in #4422.

matzf committed:
deps: update oapi-codegen to v2 (#4432)

The oapi-codegen libraries went through a reorganization; the
repositories for the oapi-codegen build tool was separated from the
runtime libraries.
In our normal code dependencies, the oapi-codegen build tool no longer
shows up. Import it in a dummy .go file as an easy way to ensure that we
have all the appropriate dependencies (via the `go.mod`/`go_deps.bzl`) .
This leaves some clutter in the `go.mod` file which is not aesthetically
pleasing, but is not a real issue due to dependency graph 
pruning.
The alternative solution would have been to manage the dependencies of
the `rules_openapi` bazel rules explicitly in a `rules_openapi_dependencies` 
macro. This turns out to be rather cumbersome, and can lead to
unexpected or even invalid dependency version combinations, because
`rules_go` does not have a way to separate the dependencies of the build
tools from the dependencies of the code.

Aside; fix `cfg = "exec"` for the oapi-codegen buildtool so that this
tool is built for the environment of the builder, not the target
platform.

matzf committed:
tools: specify project name in docker-compose files (#4396)

Simplify the usage of the various docker-compose configurations by including the project name in the configuration file. This has been supported for a while now in docker-compose v2. This allows to drop the `-p`/`--project-name` from all `docker-compose` incantations.

Also, streamline the docker-compose files generated by the topogen scripts; remove the explicit `container_name` configurations and drop all the explicit `scion_` prefixes -- managing these prefixes is docker-compose's job. 
Shut up the warnings on "SCION_EXPERIMENTAL_... variable is not set. Defaulting to a blank string." by using the variable expansion syntax to explicitly default to a blank string.

Also, drop the `docker compose` `--compatibility` flag. The compatibility flag affects what word separator is used in the
container name and it has long been deprecated. We don't usually rely on specific container names, so this should be
fine. In some exception cases where expecting specific container names seems more practical (containers for bazel-remote-cache and go-module-proxy) due to special casing in the CI scripts, container_name is set explicitly.

rohrerj committed:
fetch drkey secret value from control service

jiceathome committed:
build: make git-version not fail if no tags exist (#4438)

--always, will substitute an abbreviated commit hash if there's no tag in sight. Better than failing with a cryptic message.

jeltevanbommel committed:
Tests and MPLS maps

rohrerj committed:
prepared fetcher for future fabrid policy fetching

jiceathome committed:
testing: add router benchmark program and use it as integration test (#4437)

* Enhance the end2end_integration programs as follows:
  * end2end_integration has a slightly finer-grain role-based pair selection capability.
  * end2end_integration outputs an extra result line that provides precise start/stop timestamps for the whole test.
* process_metrics exposes the number of cores used by Go.
* Add  end2endblast. Similar to end2end but plays a game other than pingpong: packetblast. It keeps logging and tracing to a minimum. The server side responds to very few of the pings (1/256). Just enough to prove that the circuit works.
* Add an integration test called router_benchmark:
  * relies on an especially crafted topology: router_bm.topo
  * uses end2end_integration with command end2endblast over that topology as a load test for the router
  * extracts and logs relevant performance metrics after the test
  * compares performances with a set of expectations if executed by CI.

As a benchmark program, this is only a first approximation. Extending the end2end suite provided a quick solution but isn't ideal: it uses too many cores at once due to requiring 5 routers actively running for one of the tests. More work is in progress to replace this with some leaner setup where only one router is actually running.

rohrerj committed:
updated fabrid hbh header

jeltevanbommel committed:
End host policy selection using regexes (hacky, WIP, No unit-tests yet)

jeltevanbommel committed:
Sandwiching the HBH inbetween

jeltevanbommel committed:
Merge remote-tracking branch 'justin/fabrid/hbh-extension' into jvanbommel/fabrid

jeltevanbommel committed:
A crude working ping

jeltevanbommel committed:
A crude working ping

rohrerj committed:
fixed router

mlegner committed:
daemon: fix docstring for latencies in protobuf file (#4443)

Update the docstring to follow the semantics of negative means undefined.

rohrerj committed:
snet implementation for FABRID with hardcoded 0 policyID

rohrerj committed:
fixed router unit test

rohrerj committed:
add support for not specifying local ip when using scion ping with fabrid

matzf committed:
scion.sh: block for child processes at shutdown (#4445)

The `supervisorctl shutdown` command does not block for the supervisor
or its child processes to terminate.
This can occasionally lead to situations where SCION processes were
still running after the `scion.sh stop` command returned. In the CI
system, which immediately proceeds to bundle up the log files, this led
to `tar` reporting an error that the log file was still being written
to.
Fixed by invoking `supervisorctl stop all`, which does block, to
terminate all child processes before `shutdown`.

jeltevanbommel committed:
A crude working ping

jeltevanbommel committed:
Merge remote-tracking branch 'justin/fabrid/hbh-extension' into jvanbommel/fabrid

# Conflicts:
#	pkg/snet/packet.go
#	pkg/snet/path.go
#	router/dataplane.go

mlegner committed:
processmetrics: explicit type conversion to support arm64 (#4446)

rohrerj committed:
fabrid crypto bugfix

rohrerj committed:
bugfix

mlegner committed:
bazel: update bazelisk and support arm64 (#4447)

Bazel and bazilisk do support ARM64 by now. Lift the restriction on AMD64 and, while we're at it, update bazelisk to the latest version.

mlegner committed:
build: allow building without scion-ci (#4449)

The default target remains unchanged but is renamed to `build-dev`.
The target `build` now only builds the main SCION services.

jeltevanbommel committed:
wip

jeltevanbommel committed:
Ability to not use policies, as well as keep index 0 reserved.

jeltevanbommel committed:
Demo Setup

jeltevanbommel committed:
Merge remote-tracking branch 'justin/fabrid/hbh-extension' into jvanbommel/fabrid

# Conflicts:
#	pkg/experimental/fabrid/fabrid_crypto.go
#	pkg/snet/path/fabrid.go
#	router/dataplane.go

sustrik committed:
prometheus: fix copying of metrics labels (#4450)

When copying metric labels we assumed that `append` will always
create a copy of the label array. This is not necessarily the case.
In such case two metrics may end up with the same underlaying
array of labels and change to one of them also overwrites the
labels in the other one.

jeltevanbommel committed:
fix merge

rohrerj committed:
removed verify path logic of fabrid

rohrerj committed:
daemon: error out when using drkey endpoints without configured drkey (#4453)

I found a bug in the SCION daemon. If you create a SCION daemon connection and call one of the drkey endpoints without first setting the configuration value "drkey_level2_db.Connection" to something meaningful, the SCION daemon will encounter an error like that and crash:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xb82d33]
```
This bug can easily be reproduced by running a code snippet like this for a SCION daemon that has no drkey configured:
```
sd, _ := daemon.NewService(daemonAddr).Connect(ctx)
sd.DRKeyGetASHostKey(ctx, drkey.ASHostMeta{})
```
The problem is that in the file "daemon/internal/servers/grpc.go", when the drkey endpoints are called, DaemonServer.DRKeyClient is nil if it is not configured as described above.

Fixed by explicitly checking whether DRKey is available and erroring out if it is not.

lukedirtwalker committed:
pkg/log: fix panic when using custom Logger (#4452)

log.FromCtx could panic if a custom Logger was used and a span was
present in the context. This commit allows the custom logger to
implement the `WithOptions(...zap.Option) Logger` method so that the
CallerSkip can still be applied. In case the logger can't be casted to
anything the caller skip is not applied, but we also don't panic
anymore.

jiceathome committed:
testing: an improved router benchmark that only runs one router (#4444)

This is an alternative to router_benchmark and meant to eventually replace it.
It's main feature is that it only runs a single router. As a result it's measured performance is far less dependent on the number of cores available on the test system.

It is heavily inspired by the braccept test, in that:
* it relies on the test harness to create an isolated network and veth interfaces to connect the test driver to the router.
* it manufactures packets that are fed directly into the router's interfaces and it optionally captures packets for verification in the same manner.
* it assumes a custom topology and cannot be usefully run on any other topology.

It differs from braccept in that:
* It does far less verification of the forwarded packets (it's braccept's job to do that).
* It is considerably simplified, mainly because of the lighter verifications.
* It relies on a simple address-assignment scheme so code maintainers don't have to keep tens of addresses in mind.
* It does not assume that the router runs in a container network; it could also talk to a real router if given the names and mac addresses of the relevant interfaces. The test harness (or the user) is responsible for supplying the interface names and mac addresses to the test driver.
* It does not require the test harness (or the user) to have any understanding of the topology (only to configure the router with it). The specifics of the captive network to be configured are supplied by the test driver.
* The test harness doesn't need the "pause" container anymore.

The similarity between this and braccept means that, in the future, we could re-converge them. Notably, the benefits of the simple addressing scheme would make the braccept test cases easier to maintain or expand.

Collateral: the go version required was bumped to  1.21

Fixes #4442

jiceathome committed:
router: deleting the old router_benchmark integration test. (#4454)

This test has been superceeded by the router_newbenchmark test.
One component of it is preserved for its usefulness as an experiment/debugging
tool: the end2endblast component of the end2end suite.

rohrerj committed:
updated header design and fix compile error of tests

jiceathome committed:
testing: rename "router_newbenchmark" to "router_benchmark" (#4455)

Since, the original "router_benchmark" has been deleted the "new" prefix is
no longer necessary.

rohrerj committed:
query control service instead of scion daemon for drkey

jiceathome committed:
router: make the benchmark and the router more adaptive to core count (#4456)

In the router. Compute a better default for the number of processors based on multiple observations: given enough streams to occupy all processors, performance is degraded if there are as many processors as cores. Best performance is obtained by setting aside two cores for other tasks. Making this the default simplifies the benchmarking process by not having to configure an override for each type of target machine being tested.

In the benchmark. Run the test with many streams. It reflects reality and allows a very parallel router to perform at its best without distorting the performance of a less-parallel one. Make sure that the number of streams is likely to be a multiple of the number of cores so the cores are loaded evenly in spite of there being typically fewer streams than in real life. (TBD, learn the router's internal processors count and just use that).

As a side effect:
* Adjust CI's expectations for the benchmark
* Add automatic retries (2) for failed CI tests.
* Allow retrying successful CI tests (helps investigating performance variations).
* Add cpu_info to the router_benchmark output (same reason).

jiceathome committed:
testing: improved stream synthesis and cores use in router benchmark (#4457)

* testing: improved stream synthesis in router benchmark

Changes:
* Update the flowID of packets (and a couple of checksums) on the fly (just before
  sending) as opposed to pre-constructing a different packet for each stream.
* As a result, the test cases are simplified, they no-longer need to provide one
  packet per flowID.
* In passing, deleted the dumping of /proc/cpuinfo in the log. That has served its purpose.
* In the test harness, left in the form of comments, a suggestion of core partitioning. Not
  sure how to apply that in general. This is what provides the best results on my machine.

* testing: simplify the checksum updates out of the blaster.

Also enable core-partitioning to see how it behaves on the CI system.

* testing: Add check for proper saturation in the router benchmark.

In passing: fixed a couple of function names (camel-> snake)

* testing: fix expectation for droppage in the router benchmark

Was expressed in pkts instead of %.

* testing: cosmetic change to the output of expected drop ratio.

* testing: linter friendship

* testing: include a cpu selection in the router benchmark

Also: fixed the broken prom query string.

* testing: please the linter.

* testing: log lscpus on the CI system.

* testing: fix bad string formating.

* testing: Improve core selection in the router benchmark.

Added L2 cache criteria. The test harness will prefer configuration where either cores share no L2 cache or all cores share the same one.

* testing: make the router benchmark test harness compatible with more variants
of lscpu (hopefully).

* testing: log the CPU serial number in the router benchmark.

This is done as an attempt at detecting live migration.

* build: added dmidecode as a required package.

* build: Complete the list of files that affect the list of packages to install.

Three files were missing, most notably the pkgs.txt files which are typically the ones
modified when adding a dependency. As a result new packages weren't getting installed.

* build: Delete use of dmidecode.

This proved useless as a means of spotting migrations. The information doesn't
even change from instance to instance.

* build: fix trailing blank line.

* router: reduce the benchmarks expectations for saturation.

After fixing an arithmetic error, it seems we are not able to reliably cause the
router to drop more than 1% of the packets (and then sometimes no even that).
Until we figure out how to reach higher levels, making sure that other PRs
aren't blocked.

* testing: fixed bugs in router_benchmark core selection code.

* testing: update the expectations of the router benchmark.

After switching to non-overbooked aws instances, performance has become much better
and repeatable.

matzf committed:
dist: build debian packages for multiple platforms (#4448)

Build debian packages for amd64, arm64, i386 and armel.
- There are separate packages for router, control, daemon, dispatcher,
gateway and tools (scion and scion-pki).
- The packages include systemd unit files to run the services. For the
daemon, dispatcher and gateway one instance per host is supported by the
systemd service, and a default configuration file is included. For
router and control, multiple instances per host are supported, and as a
consequence of this, no default configuration file is provided.
- Currently, there is no man page contained in the packages. We should
be able to build these from our existing manuals, but it seems to
require a significant amount of fiddling to get something useful.

Building the .deb packages uses bazel with `rules_pkg`.
The target `//dist:deb_all` cross-builds packages for the default set of
target platforms. Alternatively, the target `//dist:deb` allows to build
(all) packages for the current target platform. This current platform
can be set with the `--platforms` bazel option (see
https://github.com/bazelbuild/rules_go#how-do-i-cross-compile for more
details).
- To increase reuse of build results while cross-building, some internal
targets related to openapi forcibly ignore the target platform
- The package version is based on the current git tag. `rules_pkg` can
include this version _in_ the package metadata, but bazel _cannot_ spit
out appropriately named package files. This is addressed by copying and
renaming the package files after build in a make target `make dist-deb`.

Add installation documentation for the packages and the systemd units.
As a side effect, slightly reorganize the build documentation also, trying to 
make give a simpler path to just build the binaries without installing the
entire development setup.

marcodermatt committed:
tools: fix wireshark dissector TLV options without data (#4458)

Dissecting TLV options crashes the wireshark plugin, if the `data` field
is empty. This happens always for Pad1 options and for PadN options if
`data_len` is 0.

matzf committed:
doc: update buildkite instance type (#4459)

Documenting that we've bumped the buildkite instance type from t3 to
m6i.
The main reason for the switch is that the "burstable" nature of the
t*-instances makes them extremely unsuitable for running benchmarks. The
m6i are also much faster, but not much more expensive.
Also, the agent pool is now at most 6 agents. Spinning up 10 agents and
having to warm up all the caches, and then scaling back in after a
single build, seemed excessive.

oncilla committed:
control: pluggable segment signer (#4460)

Make the segment signer pluggable. This decouples the beacon extender
from the trust signer implementation and allows plugging in different
implementations.

rohrerj committed:
added support for ingress and ip prefix defined fabrid policies in the router

jiceathome committed:
doc: update the port table to reflect current state (#4462)

Reflect the changes introduced in #4376

rohrerj committed:
updated FABRID ip prefix selection

sustrik committed:
gateway: etags added to prefix discovery protobuf (#4461)

Etags can be used so as not to transfer large prefix sets over and over
again.

Etags may or may not be supported by a particular gateway
implementation. This change is fully forward and backward compatible.
Implementations that do not support etags should just ignore the field.

Commit:
FABRID end2end initial

copy of tools/end2end, enable progress for end2end_integration.

Commit:
FABRID end2end with zero policy

- Enable zero policy in dataplane
- Fill zero policies for FABRID path if none provided
Currently fails because of mismatched ingress/egress fields.

Commit:
FABRID hop field rework

- only have one hop field for crossovers
- full end2end testing with zero policy

Commit:
Check for peering assumption

Ensure that peering paths only have two segments.

Commit:
End2End path validation

- Computation of path validation at source
- Fix HostHostKey fetching at source, flipped source/destination
- Derivation of path key at destination
- Verification of path validator at destination

rohrerj committed:
Merge pull request #8 from marcodermatt/fabrid/end2end-integration

Fabrid/end2end integration

rohrerj committed:
updated FABRID router implementation and added mpls unit tests

rohrerj committed:
reorder imports

rohrerj committed:
implemented fabrid interface map and default values

rohrerj committed:
added unit tests

jeltevanbommel committed:
- MPLS labels can now be assigned per connection point.
- MPLS labels for IP ranges added and synced with BR
- Introduce explicit wildcard ingress interface in FABRID

jeltevanbommel committed:
Changing the FabridPolicyPerHop to FabridPolicyIdentifier for better readability + it didn't work before

jeltevanbommel committed:
Removed the sequence parameter to provide fabrid policies, replaced with a different path constraints mechanism. Generic code cleanup, refactored such that there are not multiple different policy types in different packages anymore, for a bit more consistency.

jeltevanbommel committed:
Small rename

jeltevanbommel committed:
Merge remote-tracking branch 'refs/remotes/scionlab/scionlab' into fabrid/base-implementation

# Conflicts:
#	BUILD.bazel
#	daemon/internal/servers/grpc.go
#	pkg/daemon/grpc.go
#	pkg/proto/control_plane/seg_extensions.pb.go
#	pkg/proto/daemon/daemon.pb.go
#	pkg/snet/path.go
#	private/path/combinator/graph.go
#	proto/daemon/v1/daemon.proto
#	router/dataplane.go
#	tools/dc

jeltevanbommel committed:
Fixed issue with policy ids not getting correctly set.

jeltevanbommel committed:
Added wireshark support for fabrid extensions.

jeltevanbommel committed:
Regenerate antlr

jeltevanbommel committed:
Regenerate antlr

rohrerj committed:
FABRID design document

Created the Design Document for the SCIONLab FABRID implementation.

jeltevanbommel committed:
Cleanup and reuse of already created hops datastructure in snet/fabrid, and added back the explicit zero index policy.

jeltevanbommel committed:
Cleanup and reuse of already created hops datastructure in snet/fabrid, and added back the explicit zero index policy.

rohrerj commented on tools/wireshark/scion.lua:
```diff
@@ -504,6 +504,8 @@ local scion_extn_tlv_option_types = {
   [0] = "Pad1",
   [1] = "PadN",
   [2] = "Packet Authenticator Option",
+  [3] = "Fabrid Packet Identifier",
```
I would not name it Fabrid Packet Identifier because the idea is to use this also with other extensions. Maybe just something like "Packet Identifier"

rohrerj commented on tools/fabrid/main.go:
```diff
@@ -0,0 +1,683 @@
+// Copyright 2018 ETH Zurich
```
Since this file is added and not modified, the copyright part does not make sense. Probably a copy-paste mistake

jeltevanbommel replied:
This was Marc's work, maybe he's a time traveler ;)
fixed

rohrerj commented on FabridTopo.png:
Does this really belong into the repository?

jeltevanbommel replied:
At some point it did, now no more, so removed.

rohrerj commented on control/beaconing/extender.go:
```diff
@@ -17,9 +17,11 @@ package beaconing
 import (
 	"context"
 	"encoding/binary"
+	fabridext "github.com/scionproto/scion/pkg/segment/extensions/fabrid"
```
Move import down to the other scionproto imports

rohrerj commented on control/config/fabrid.go:
```diff
@@ -0,0 +1,103 @@
+package config
+
+import (
+	"github.com/scionproto/scion/pkg/segment/extensions/fabrid"
```
Move import down to the other scionproto imports

rohrerj commented on control/fabrid/fabrid_manager.go:
```diff
@@ -0,0 +1,154 @@
+package fabrid
+
+import (
+	"github.com/scionproto/scion/control/config"
```
Group the imports the same way as for the other files

rohrerj commented on control/config/fabrid.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// Validate validates that all values are parsable.
+func (cfg *FABRIDPolicy) Validate() error {
```
Is this function even used?

jeltevanbommel replied:
A leftover from when it was part of the toml config, it was auto called there. Added a call now

rohrerj commented on control/config/fabrid_test.go:
```diff
@@ -0,0 +1,109 @@
+package config_test
+
+import (
+	"bytes"
```
Group the imports the same way as for the other files

rohrerj commented on control/fabrid/fabrid_manager.go:
```diff
@@ -0,0 +1,154 @@
+
+const MaxFabridPolicies = 255
+
+// TODO(jvanbommel): Can probably combine this with PolicyIdentifier
```
Yes I also think so

jeltevanbommel replied:
Not doing it at the moment as there is some extra data in the policyidentifier to signal global/local policy. I think I will refactor that to just be a bool flag. After that there's less redundancy.

rohrerj commented on control/fabrid/fabrid_manager.go:
```diff
@@ -0,0 +1,154 @@
+	return nil
+}
+
+func (f *FabridManager) Active() bool {
```
At the moment this function always returns true. Either add some logic or remove it

jeltevanbommel replied:
Removed

rohrerj commented on control/fabrid/fabrid_manager.go:
```diff
@@ -0,0 +1,154 @@
+	return true
+}
+
+func (f *FabridManager) parseAndAdd(path string, fi os.FileInfo, err error) error {
```
argument err is overwritten before first use (SA4009)
If you don't care about the value please use "_" in the function parameter for error.
But based on the documentation of WalkFunc you might still have to handle this case somehow:
"The err argument reports an error related to path, signaling that Walk will not walk into that directory. The function can decide how to handle that error; as described earlier, returning the error will cause Walk to stop walking the entire tree."

jeltevanbommel replied:
I think this should just be a return nil + Done

rohrerj commented on control/fabrid/fabrid_manager.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func (f *FabridManager) parseAndAdd(path string, fi os.FileInfo, err error) error {
+	if !fi.Mode().IsRegular() {
```
Maybe add a comment that is also makes sure that the current "file" is not a directory

jeltevanbommel replied:
Added

rohrerj commented on control/fabrid/fabrid_manager.go:
```diff
@@ -0,0 +1,154 @@
+			Type: fabrid_ext.Wildcard,
+		}
+	}
+	return fabrid_ext.ConnectionPoint{}
```
Is the "else" case an intended case? If yes, write a comment for which case this is used. If not, return an error or at least check from the calling function whether the returned value is valid.

jeltevanbommel replied:
Good catch, it should error out yes. Added

rohrerj commented on control/fabrid/fabrid_manager_test.go:
```diff
@@ -0,0 +1,17 @@
+
+package fabrid
+
+//TODO(jvanbommel) Test loading policies from a folder.
```
No tests so far

jeltevanbommel replied:
:)

rohrerj commented on control/fabrid/grpc/fabrid_service.go:
```diff
@@ -0,0 +1,95 @@
+package grpc
+
+import (
+	"bytes"
```
Group imports

rohrerj commented on control/fabrid/grpc/fabrid_service.go:
```diff
@@ -0,0 +1,95 @@
+		return &experimental.PolicyDescriptionResponse{}, err
+	}
+
+	s.FabridManager.RemotePolicyCache[identifier] = fabrid.RemotePolicyDescription{Description: policy.Description, Expires: time.Now().Add(time.Hour * 3)}
```
Why do you use as validity 3 hours? Why not 2? Why now 4?
But maybe it would make sense that the remote AS can decide on the validity

jeltevanbommel replied:
Shoould definitely make it some configurable constant

jeltevanbommel replied:
Done, it's configurable in the CS now

rohrerj commented on control/fabrid/grpc/fabrid_service.go:
```diff
@@ -0,0 +1,95 @@
+func (s Server) GetRemotePolicyDescription(ctx context.Context, request *experimental.RemotePolicyDescriptionRequest) (*experimental.PolicyDescriptionResponse, error) {
+	//TODO(jvanbommel): signature / hash?
+	identifier := fabrid.RemotePolicyIdentifier{ISDAS: request.IsdAs, Identifier: request.PolicyIdentifier}
+	if val, ok := s.FabridManager.RemotePolicyCache[identifier]; ok && val.Expires.UnixNano() > time.Now().UnixNano() {
```
I would suggest to replace:
val.Expires.UnixNano() > time.Now().UnixNano()
with:
val.Expires.After(time.Now())

rohrerj commented on control/fabrid/grpc/fabrid_service.go:
```diff
@@ -0,0 +1,95 @@
+	mplsIpMap := make(map[uint32]*experimental.MPLSIPArray)
+	for i, entry := range s.FabridManager.MPLSMap.IPPoliciesMap {
+		if _, exists := mplsIpMap[i]; !exists {
+			mplsIpMap[i] = &experimental.MPLSIPArray{Entry: make([]*experimental.MPLSIP, 0)}
```
I would suggest to initialize the Entry array like that:
Entry: make([]*experimental.MPLSIP, 0, len(entry))
Otherwise the initially memory allocation for the array might be too small and when appending it has to do additional memory allocations

jeltevanbommel replied:
Good catch

rohrerj commented on control/fabrid/grpc/fabrid_service_test.go:
```diff
@@ -0,0 +1,265 @@
+package grpc
+
+import (
```
Group imports

rohrerj commented on control/fabrid/grpc/fabrid_service_test.go:
```diff
@@ -0,0 +1,265 @@
+		fabrid_ext.SupportedIndicesMapFromPB(indices.SupportedIndicesMap))
+}
+
+func TestGetIndexIdentifierMap(t *testing.T) {
```
Does not compile

rohrerj commented on control/fabrid/grpc/fetcher.go:
```diff
@@ -0,0 +1,128 @@
+
+package grpc
+
+import (
```
Group imports

rohrerj commented on control/fabrid/grpc/fetcher.go:
```diff
@@ -0,0 +1,128 @@
+		return p, nil
+	}
+	// we've tried out all the paths; we reset the map to retry them.
+	f.errorPaths = make(map[snet.PathFingerprint]struct{})
```
I think `clear(f.errorPaths)` might be more efficient than a new memory allocation

jeltevanbommel replied:
As far as I can see, in both cases you wait until a gc cycle to actually free up the memory.

rohrerj commented on control/fabrid/grpc/fetcher_test.go:
```diff
@@ -0,0 +1,95 @@
+
+package grpc_test
+
+import (
```
Group imports

rohrerj commented on control/fabrid/mpls.go:
```diff
@@ -0,0 +1,106 @@
+package fabrid
+
+import (
```
Group imports

rohrerj commented on control/fabrid/mpls.go:
```diff
@@ -0,0 +1,106 @@
+	return orderedKeys
+}
+
+//when scion is on go1.18 make it generic
```
Based on the go.mod file, aren't we already on go 1.21?

jeltevanbommel replied:
After the merge yes, added the generic replacement

rohrerj commented on control/fabrid/mpls.go:
```diff
@@ -0,0 +1,106 @@
+//}
+
+// This method is to be called after all inserts and removes from the internal map
+// TODO(jvanbommel): this feels too expensive for what a relatively simple synchronization need. Revise?
```
If I understand the code correctly, this is only called on `FabridManager.Load()` so that should not take too much time. But you could still optimize it. And do we really need to use a cryptographic hash function here?

jeltevanbommel replied:
We could just use a lastupdated unix epoch, definitely a lot more lightweight. The only issue would be an AS restarting, and reloading all policies, causing all routers to update. I've changed it to a non-cryptographic hash for now.

rohrerj commented on control/fabrid/fabrid_manager.go:
```diff
@@ -0,0 +1,154 @@
+	return fb, fb.Load()
+}
+
+func (f *FabridManager) Reload() error {
```
Is this function even used? I cannot find a function that calls Reload()

jeltevanbommel replied:
Currently unused. Is for future purposes when we do live reloading of policies etc.

rohrerj commented on fabrid.yml:
Does this really belong in the repository?

rohrerj commented on tools/topology/go.py:
```diff
@@ -114,6 +114,7 @@ def _build_control_service_conf(self, topo_id, ia, base, name, infra_elem, ca):
             'general': {
                 'id': name,
                 'config_dir': config_dir,
+                'fabrid_path': 'gen-fabrid/',
```
With this I think every AS would use the same FABRID policies which does not make sense. At least not for the local policies.

jeltevanbommel replied:
Used to be for development purposes, but yes at this point I also overwrite them manually. Revert these changes or do you want to change the topo generator to have a fabrid flag?

rohrerj commented on daemon/internal/servers/grpc.go:
```diff
@@ -174,21 +179,46 @@ func pathToPB(path snet.Path) *sdpb.Path {
 		Interface: &sdpb.Interface{
 			Address: &sdpb.Underlay{Address: nextHopStr},
 		},
-		Interfaces:      interfaces,
```
Spacing for variables are not alligned.

rohrerj commented on gen-fabrid/1-global_example.yml:
Does this really belong in the repository?

rohrerj commented on gen-fabrid/55-local_example.yml:
Does this really belong in the repository?

jeltevanbommel replied:
Was mostly for ease of use when there was no design doc. removed now

rohrerj commented on pkg/daemon/grpc.go:
```diff
@@ -16,12 +16,13 @@ package daemon
 import (
 	"context"
-	"net"
+		"net"
```
Spacing

rohrerj commented on pkg/daemon/grpc.go:
```diff
@@ -16,12 +16,13 @@ package daemon
 	"google.golang.org/grpc"
 	"google.golang.org/protobuf/types/known/timestamppb"
-
+	"github.com/scionproto/scion/pkg/experimental/fabrid"
```
Add a line break between the imports of google.golang.org.... and github,com/scionproto...

rohrerj commented on pkg/daemon/grpc.go:
```diff
@@ -280,16 +286,17 @@ func convertPath(p *sdpb.Path, dst addr.IA) (path.Path, error) {
+			Expiry:         expiry,
+			Latency:        latency,
+			Bandwidth:      p.Bandwidth,
 			CarbonIntensity: p.CarbonIntensity,
```
Spacing of the variables is not alligned

rohrerj commented on pkg/experimental/fabrid/crypto/fabrid_crypto.go:
```diff
@@ -0,0 +1,243 @@
+	"crypto/cipher"
+	"encoding/base64"
+	"encoding/binary"
+	"github.com/scionproto/scion/pkg/experimental/fabrid"
```
Grouping of imports is wrong

rohrerj commented on pkg/experimental/fabrid/crypto/fabrid_crypto_test.go:
```diff
@@ -0,0 +1,167 @@
+						}
+					}
+
+					err := crypto.InitValidators(f, id, s, tmpBuffer, pathKey, asHostKeys, asAsKeys, ases, ingresses, egresses)
```
Does not compile

rohrerj commented on pkg/segment/extensions/digest/digest_test.go:
```diff
@@ -15,6 +15,7 @@
 package digest_test
 
 import (
+	"github.com/scionproto/scion/pkg/segment/extensions/fabrid"
```
Import not in the correct group

rohrerj commented on pkg/segment/extensions/fabrid/fabrid.go:
```diff
@@ -0,0 +1,347 @@
+package fabrid
+
+import (
```
Imports are not grouped correctly

rohrerj commented on pkg/segment/extensions/fabrid/fabrid.go:
```diff
@@ -0,0 +1,347 @@
+	}
+}
+
+func (d *Detached) String() string {
```
Is this function even used?

jeltevanbommel replied:
Currently no, but having a string method for a complex datastructure is useful imo for debugging

rohrerj commented on pkg/segment/extensions/fabrid/fabrid_test.go:
```diff
@@ -0,0 +1,588 @@
+
+package fabrid
+
+import (
```
Group imports

rohrerj commented on pkg/segment/extensions/fabrid/fabrid_test.go:
```diff
@@ -0,0 +1,588 @@
+				SupportedIndicesMap: SupportedIndicesMap{},
+				IndexIdentiferMap: IndexIdentifierMap{
+					2: &PolicyIdentifier{
+						Type:       GlobalPolicy,
```
Does not compile

rohrerj commented on pkg/segment/unsigned.go:
```diff
@@ -15,32 +15,52 @@
 package segment
 
 import (
```
Group imports

rohrerj commented on pkg/segment/unsigned_test.go:
```diff
@@ -15,6 +15,7 @@
 package segment
 
 import (
+	"github.com/scionproto/scion/pkg/segment/extensions/fabrid"
```
Group imports

rohrerj commented on pkg/slayers/extension/fabrid.go:
```diff
@@ -0,0 +1,197 @@
+	return nil
+}
+
+// DecodeForCurrentHop uses the scion meta header to determine the current hop
```
Function description is not up to date

rohrerj replied:
Fixed

rohrerj commented on pkg/slayers/extension/fabrid_test.go:
```diff
@@ -0,0 +1,204 @@
+
+package extension_test
+
+import (
```
Group imports

rohrerj commented on pkg/slayers/extension/identifier_test.go:
```diff
@@ -0,0 +1,141 @@
+
+package extension_test
+
+import (
```
Group imports

rohrerj commented on pkg/snet/packet.go:
```diff
@@ -344,8 +348,8 @@ type Packet struct {
+		e2eLayer   slayers.EndToEndExtn
 		udpLayer   slayers.UDP
 		scmpLayer  slayers.SCMP
 	)
```
A few lines down at `decoded := make([]gopacket.LayerType, 0, 4)`, we have to make sure that this still works if we have packets that have all layers present. Maybe we have to increase the capacity from 4 to 6

rohrerj replied:
Fixed

rohrerj commented on pkg/snet/path.go:
```diff
@@ -159,16 +197,18 @@ func (pm *PathMetadata) Copy() *PathMetadata {
+		Expiry:         pm.Expiry,
+		Latency:        append(pm.Latency[:0:0], pm.Latency...),
+		Bandwidth:      append(pm.Bandwidth[:0:0], pm.Bandwidth...),
 		CarbonIntensity: append(pm.CarbonIntensity[:0:0], pm.CarbonIntensity...),
```
Spacing to alligned correctly

rohrerj commented on pkg/snet/path/fabrid.go:
```diff
@@ -0,0 +1,232 @@
+import (
+	"context"
+	"fmt"
+	"github.com/scionproto/scion/pkg/experimental/fabrid/crypto"
```
Group imports

rohrerj commented on pkg/snet/path/fabrid.go:
```diff
@@ -0,0 +1,232 @@
+	return f, nil
+}
+
+func (f *FABRID) RegisterDRKeyFetcher(fn func(context.Context, drkey.ASHostMeta) (drkey.ASHostKey, error),
```
Maybe we can implement this a bit better. I think the reason why I implemented it like that was because I was not able to import the functions directly because of recursive dependencies.

rohrerj reviewed (commented):
Partial review

jeltevanbommel committed:
Reformatted and passes linter

jeltevanbommel committed:
Cleaned up repo

jeltevanbommel committed:
Configurable remote cache duration and removed Active() method

jeltevanbommel commented on control/config/fabrid.go:
```diff
@@ -0,0 +1,105 @@
+	"github.com/scionproto/scion/private/config"
+)
+
+type FABRIDPolicy struct {
```
I'll change this so there isn't both a LocalIdentifier and GlobalIdentifier. Also rename to FabridPolicyDefinition, so it's a bit less confusing with fabrid.Policy

jeltevanbommel committed:
WIP, test for fabrid_manager, fixed other tests, as well as small refactorings for fabrid manager and snet/path

jeltevanbommel committed:
WIP, test for fabrid_manager, fixed other tests, as well as small refactorings for fabrid manager and snet/path

rohrerj committed:
use scion daemon for fabrid drkey

rohrerj committed:
updated drkey fetcher of the router

rohrerj committed:
bugfix of fetcher

jeltevanbommel committed:
WIP, test for fabrid_manager, fixed other tests, as well as small refactorings for fabrid manager and snet/path

marcodermatt committed:
fix: DRKeyProvider initialization race condition

On ./scion.sh start, some BRs failed with SPAWN ERROR, because SecretKeyFetcher tried to access the DRKeyProvider mutex before it was initialized by DataPlane.Run().

marcodermatt committed:
refactor: Add --fabrid flag to end2end integration

Remove separate fabrid end2end executable and integrate with existing one. To test it, start the topology and run `./bin/end2end_integration --fabrid`. Currently requires the `validate*Req` functions in drkey_service.go to be disabled, as the request IPs don't match.
Fix: Dataplane used timestamp of the current InfoField for parsing the Identifier option, but it should always use the first InfoField.

Commit:
refactor: Replace policy global/local type with flag

marcodermatt committed:
Add fabrid flag to border router config

Commit:
Merge dispatcherless branch into fabrid/base

Significant changes in:
- dataplane test for adding interfaces
- end2end packet processing. Still requires a fix for server side path validation.

rohrerj committed:
changed scion daemon dialer for drkey

rohrerj committed:
temporary solution to use scion daemon address as local ip in end2end integration test

rohrerj committed:
changed scionping to use daemon addr as local addr in the fabrid case if unset

Commit:
Add --fabrid flag to scion.sh topology

Specifying the fabrid flag enables DRKey and FABRID for all ASes on CS, SD and BRs.

Commit:
Use --fabridquery as enable flag in ping

rohrerj committed:
updated expected values for segment unit tests

Commit:
Show FABRID status in path requests

Extract FABRID capability from beacons, based on presence of FABRID digest. Propagate to SD and endhosts, allowing them to use zero policy on ASes without any configured polices.

rohrerj committed:
fixed border router benchmark acceptance test

rohrerj commented on control/config/fabrid.go:
```diff
@@ -0,0 +1,104 @@
+		return serrors.New("Invalid IPv6 Address range for connection point")
+	} else if cfg.Type == fabrid.IPv4Range && (cfg.IPAddress == "" || cfg.Prefix > 32) {
+		return serrors.New("Invalid IPv4 Address range for connection point")
+	}
```
I think it might make sense to check whether cfg.IPAddress can be parsed to an actual IP Address or not

rohrerj replied:
Fixed

rohrerj commented on control/config/fabrid.go:
```diff
@@ -0,0 +1,104 @@
+	} else if cfg.Ingress.Type == fabrid.Interface && cfg.Egress.Type == fabrid.Interface && cfg.
+		Ingress.Interface == cfg.Egress.Interface {
+		return serrors.New("Interfaces should be distinct")
+	}
```
Maybe check whether the ingress / egress even exist / are valid

rohrerj replied:
Fixed

rohrerj committed:
added some error tests for fabrid crypto

Commit:
Use PacketConn for FABRID end2end integration

Performs path validation at destination for end2end FABRID test. Requires patching drkey_service to allow servers to fetch DRKeys.

rohrerj committed:
updated function description of DecodeForHF

rohrerj committed:
changed capacity of decoding layers from 4 to 6

rohrerj commented on pkg/snet/path/fabrid.go:
```diff
@@ -0,0 +1,216 @@
+	}
+	//TODO(jvanbommel) How was this remove hops introduced by crossovers even supposed to work?
+	// Decreasing the numHops would still add the hops to ingresses and egresses of crossovers
+	// in the middle of the path, and even cut out the end of the path. What did I miss?
```
Is this resolved?

jeltevanbommel replied:
@marcodermatt

marcodermatt replied:
Resolved by only passing FABRID relevant hops, instead of the two hopfields SCION uses for a crossover or some intermediary hops.
Comment can be removed

rohrerj commented on private/path/fabridquery/matchlist.go:
```diff
@@ -0,0 +1,73 @@
+	SelectedPolicies []*Policy
+}
+
+// Copy creates a deep copy of the MatchList object, including the list of selected policies.
```
Are you sure that this makes a deep copy? For me this looks like it just copies the pointer array but not the policies itself.

rohrerj replied:
Removed "deep"

rohrerj commented on tools/integration/integrationlib/common.go:
```diff
@@ -62,6 +62,33 @@ func Setup() error {
 	validateFlags()
+	// this is a temporary solution to make sure that the local IP = daemon IP
+	// TODO: remove once a proper solution has been established
+	if Mode == ModeClient {
```
We still need a proper fix for this

rohrerj committed:
fixed concurrency bug when initializing drkey provider in router

Commit:
Use SD address for end2end FABRID test

When generating test pairs with the `--fabrid` flag set, both client and server use the same IP as the respective scion daemon. Does not yet work on the docker topology

rohrerj committed:
added checks in control service when loading fabrid policies to check whether provided interfaces exist and whether provided ip address strings are actually ip addresses

rohrerj committed:
changed code comment from deep copy to copy

Commit:
Only do path validation if last hop supports FABRID

The pathKey variable is set to nil, if the last hop does not support FABRID, disabling fetching of the HostHostKey and computation of the path validator.

Commit:
Passes linter

rohrerj committed:
docker setup for fabrid

Commit:
linting

Commit:
Merge branch 'refs/heads/scionlab' into fabrid/base

# Conflicts:
#	daemon/internal/servers/grpc.go
#	doc/dev/design/FABRID.rst
#	pkg/daemon/grpc.go
#	pkg/private/xtest/graph/graph.go
#	pkg/proto/control_plane/seg_extensions.pb.go
#	pkg/proto/daemon/daemon.pb.go
#	pkg/snet/path.go
#	private/path/combinator/graph.go
#	proto/control_plane/v1/seg_extensions.proto
#	proto/daemon/v1/daemon.proto
